lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -566,8 +641,45 @@ class HandshakeConversionPattern : public OpConversionPattern<T> {
     setAllReadyWithCond(s, inputs, output, allValid);
   }
 
-  void buildMuxLogic(RTLBuilder &s, ArrayRef<InputHandshake> inputs,
-                     InputHandshake &cond, OutputHandshake &output) const {}
+  // Builds mux logic for the given inputs and outputs.
+  // Note: it is assumed that the caller has removed the 'select' signal from
+  // the 'unwrapped' inputs and provide it as a separate argument.
+  void buildMuxLogic(RTLBuilder &s, UnwrappedIO &unwrapped,
+                     InputHandshake &select) const {
+    // ============================= Control logic =============================
+    // Decimal-to-1-hot decoder. 'shl' operands must be identical in size.
+    size_t size = unwrapped.inputs.size();
+    Value truncatedSelect = select.data.getType().getIntOrFloatBitWidth() > size
+                                ? s.truncate(select.data, size)
+                                : select.data;
+    auto select1h = s.shl(s.constant(size, 1), s.zext(truncatedSelect, size));
```

Comment:
I don't know if there's any issue in this particular instance, but in general, I remember the EDSC folks got burned by C++ argument evaluation order being unspecified. So I think you could get something like

```mlir
%0 = hw.constant ...
%1 = hw.zext ...
%2 = hw.shl ...
```

From one compiler, but 

```
%0 = hw.zext ...
%1 = hw.constant ...
%2 = hw.shl ...
```

From another. I don't know if there will be major issues here, but at the least it is relying on undefined behavior, and that could make reproducibility harder (at least, the tests might need to all use CHECK-DAG, for example).

I think the safest thing is to still follow the same rules as we would for a normal `builder.create` call: pull out a temporary for one or the other argument, so no more than one argument to each function call is creating IR.